### DISCUSSION TITLE
Can't build on machines with too little RAM

### DISCUSSION THREAD

**Discussion by nerdhardest** [uutils/coreutils#4714] [General]
I found that 'cargo build --release' fails at the final step when
the machine on which I was building Rust coreutils had too
little RAM.

I found that I needed 3-4 GB of RAM to build the entire
'busybox'-style coreutils.  (Slightly less on a 32-bit x686 machine.)

I was, however, able to build smaller individual utils, on these
smaller machines.

I found that the entire Rust coreutils produced a binary of ~10 MB,
while some of the individual binaries were ~2.5 MB or less (which
is still pretty big for trivial utilities).

**Comment by tertsdiepraam** [comment]
The first point we can't do much about. Maybe it's the LTO stage? In that case you could try settings `lto = false` in the `Cargo.toml`. If that doesn't help, you could download the prebuilt binaries though or build on another machine and copy the binaries.

The binaries are indeed pretty big by default. This is by design, because we include debug info. The assumption there is that people who build the util via cargo are usually the ones who also want to contribute bug reports. You can make it much smaller though, by using the `release-fast` and `release-small` profiles. Compared to GNU, we use static linking, so there will always be a bit of a difference.

Hope that helps!

**Reply by nerdhardest** [reply]
Setting 'lto = false' in Cargo.toml did the trick!

I was able to build Rust coreutils on a 64-bit Raspberry Pi 4 w/2 GB RAM with lto=false.

However, lto=false produces a binary which is 50% bigger !

I then tried --profile=release-small and I now get a binary which is ~5MB, which
is about *half* the size of a 'release' binary.

Building uu_factor alone with --profile=release-small produces a binary
of ~600K, whereas the binary with --profile=release is ~2.5MB.

**Reply by tertsdiepraam** [reply]
> However, lto=false produces a binary which is 50% bigger !

Yep, that's why we enabled it by default. It's all trade-offs unfortunately.

**Reply by nerdhardest** [reply]
I can confirm that turning 'lto' off allows building Rust coreutils on almost all of my
memory-constrained Linux systems -- even machines with 1 GB RAM.

With the 'release-small' profile, the memory footprint with and without 'lto' are
about the same (on machines that work with 'lto' on); I don't know about running
speed comparisons, however.

I also turned off 'incremental', because it doesn't seem to help very much and I'm
not certain that I can trust it completely.

**Reply by nerdhardest** [reply]
On my original 32-bit Raspberry Pi Model B (rev 2) with 0.5 GB RAM,
building Rust coreutils with lto=false, incremental=false, profile is release-small,
takes *2 hours and 16 minutes*, but it *does* finish successfully !!
